examples/hosts/hosts-sample/src/codeDetailsLoader.ts

```diff
@@ -0,0 +1,107 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import * as semver from "semver";
+import {
+    IFluidCodeDetails,
+    IFluidCodeDetailsComparer,
+} from "@fluidframework/core-interfaces";
+import {
+    ICodeDetailsLoader,
+    IFluidModuleWithDetails,
+} from "@fluidframework/container-loader";
+import { fluidExport } from "./fauxComponent";
+import { parsePackageDetails } from "./utils";
+
+/**
+ * Emulates a code loader capable of resolving Fluid modules defined in a local in-memory store
+ * given the package name and version. The loader's implementation showcases one of possible
+ * strategies of managing code version upgrades.
+ */
+export const InMemoryCodeDetailsLoader = new (class
+    implements ICodeDetailsLoader, IFluidCodeDetailsComparer {
+    /**
+     * The load method is called by the Fluid loader prior to instantiating the runtime.
+     *
+     * @param source - Code details written in the quorum.
+     * @returns - Module entry point along with the module's own code details.
+     */
+    async load(source: IFluidCodeDetails): Promise<IFluidModuleWithDetails> {
+        // Get normalized package info
+        const { name: sourcePackageName } = parsePackageDetails(source.package);
+        if (sourcePackageName === undefined) {
+            // Raise an error when code details written in the document differ from
+            // the package supported by this code loader.
+            throw new Error(
+                "The document is created with a different package that is not supported by this code loader.",
+            );
+        }
+        // Assemble and return the resolved Fluid module accompanied by the code details characterizing
+        // its capabilities, such as supported range of code versions.
+        // In this scenario we use the version 2.0.0 as the latest available in the store.
+        return {
+            module: { fluidExport },
+            details: {
+                package: {
+                    name: sourcePackageName,
+                    version: "2.0.0",
```

Comment:
No, it breaks the app as I described earlier
> And the problem is the code upgrade proposal is not saved in the summary snapshot as the container gets closed immediately after the proposal submitted. As a result, it's not possible to open the document as it will keep failing.

Reply:
This seems problematic -- I imagine the mainstream use case for most of our customers will generally be something like

1. Load on 1.0.0
2. Propose 1.1.0, 2.0.0, etc.
3. Either keep running (if `satisfies`) or safely reload on the upgraded version

I'll approve this PR just so there's some baseline to work from, but if we don't have a solution for this flow then I think we need to design one.

Reply:
I agree. The whole code upgrade mechanism looks half-baked now and requires more investment to make it work. Thanks!